graph/graph.go

```diff
@@ -402,6 +404,43 @@ func (graph *Graph) ByParent() (map[string][]*image.Image, error) {
 	return byParent, err
 }
 
+// If the images and layers are in pulling chain, mark them.
+// If not, they may be deleted by rmi with dangling condition.
+func (graph *Graph) AddtoPulling(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.pullingImgs[layerID]; exists {
+			graph.pullingImgs[layerID] = append(imgs, imgID)
+		} else {
+			graph.pullingImgs[layerID] = []string{imgID}
+		}
+	}
+}
+
+// If the images and layers have benn pulled, unmark them.
+func (graph *Graph) DelfromPulling(layerIDs []string, imgID string) {
+	for _, layerID := range layerIDs {
+		if imgs, exists := graph.pullingImgs[layerID]; exists {
+			result := make([]string, 0)
+			index := 0
+			endIndex := len(imgs) - 1
+			for k, v := range imgs {
+				if v == imgID {
+					result = append(result, imgs[index:k]...)
+					index = k + 1
+				} else if k == endIndex {
+					result = append(result, imgs[index:endIndex+1]...)
+				}
+			}
+			if current := len(result); current == 0 {
+				delete(graph.pullingImgs, layerID)
+			} else {
+				imgs = make([]string, len(result))
```

Comment:
This will not update the value in the map if the new result overflows the memory in imgs. A new slice will be allocated and the map value will not be updated. This needs to be something like this:

```
copy(graph.pullingImgs[layerID], result)
```

EIther way, I am not sure that slice is the right datastructure here.